x/pairing/keeper/msg_server_relay_payment.go

```diff
@@ -48,91 +49,102 @@ func (k msgServer) RelayPayment(goCtx context.Context, msg *types.MsgRelayPaymen
 		badgeSig := []byte{}
 		// if badge is found in the map, clientAddr will change (assuming the badge is valid) since the badge user is not a valid consumer (the badge signer is)
 		if badgeFound {
-			if !badgeData.Badge.IsBadgeValid(clientAddr.String(), relay.LavaChainId, uint64(relay.Epoch)) {
-				return nil, utils.LavaFormatWarning("badge must match traits in relay request", fmt.Errorf("invalid badge"),
-					utils.Attribute{Key: "badgeAddress", Value: badgeData.Badge.Address},
-					utils.Attribute{Key: "badgeLavaChainId", Value: badgeData.Badge.LavaChainId},
-					utils.Attribute{Key: "badgeEpoch", Value: badgeData.Badge.Epoch},
-					utils.Attribute{Key: "relayAddress", Value: clientAddr},
-					utils.Attribute{Key: "relayLavaChainId", Value: relay.LavaChainId},
-					utils.Attribute{Key: "relayEpoch", Value: relay.Epoch},
-				)
-			}
-
-			badgeUsedCuKey := types.BadgeUsedCuKey(badgeData.Badge.ProjectSig, relay.Provider)
-			badgeUsedCuMapEntry, found := k.GetBadgeUsedCu(ctx, badgeUsedCuKey)
-			if !found {
-				// setting a timer with a callback to delete the badgeUsedCuEntry after badge.Epoch+blocksToSave (see keeper.go)
-				badgeUsedCuTimerExpiryBlock := k.BadgeUsedCuExpiry(ctx, badgeData.Badge)
-				if badgeUsedCuTimerExpiryBlock <= uint64(ctx.BlockHeight()) {
-					return nil, utils.LavaFormatWarning("badge rejected", fmt.Errorf("badge used CU entry validity expired"),
-						utils.Attribute{Key: "badgeUsedCuTimerExpiryBlock", Value: badgeUsedCuTimerExpiryBlock},
-						utils.Attribute{Key: "currentBlock", Value: uint64(ctx.BlockHeight())},
-					)
-				}
-				// timerKey = badgeUsedCuMapKey since all badgeUsedCuMapKey keys are unique - can be used to differentiate the timers
-				k.badgeTimerStore.AddTimerByBlockHeight(ctx, badgeUsedCuTimerExpiryBlock, badgeUsedCuKey, []byte{})
-
-				badgeUsedCuMapEntry = types.BadgeUsedCu{
-					BadgeUsedCuKey: badgeUsedCuKey,
-					UsedCu:         0,
-				}
-			}
-
-			// enforce badge CU overuse
-			if relay.CuSum+badgeUsedCuMapEntry.UsedCu > badgeData.Badge.CuAllocation {
-				return nil, utils.LavaFormatWarning("badge CU allocation exceeded", fmt.Errorf("could not update badge's used CU"),
-					utils.Attribute{Key: "relayCuSum", Value: relay.CuSum},
-					utils.Attribute{Key: "badgeCuLeft", Value: badgeData.Badge.CuAllocation - badgeUsedCuMapEntry.UsedCu},
-				)
+			newBadgeTimerExpiry, err = k.checkBadge(ctx, badgeData, clientAddr.String(), relay)
+			if err != nil {
+				utils.LavaFormatWarning("badge check failed", err)
+				rejectedCu += relay.CuSum
+				continue
 			}
 
-			badgeUsedCuMapEntry.UsedCu += relay.CuSum
-			k.SetBadgeUsedCu(ctx, badgeUsedCuMapEntry)
-
 			// badge is valid & CU enforced -> switch address to badge signer (developer key) and continue with payment
 			clientAddr = badgeData.BadgeSigner
 			badgeSig = badgeData.Badge.ProjectSig
 		}
 
 		providerAddr, err := sdk.AccAddressFromBech32(relay.Provider)
```

Comment:
the provider address is always == creator, if this check fails you can fail all the tx since all the relays come from the same provider

Reply:
done. see https://github.com/lavanet/lava/commit/1090981f67cb42d98190a484b5bbfc0bd645c78e